core: Add parser for more complex pass pipeline specs

AntonLydike opened the pull request:
Based on top of #955 

This adds a parser for pass pipelines. This parser works as follows:

```python
list(
    parse_pipeline(
        'pass-1,pass-2{arg1=1 arg2=test arg3="test-str" arg-4=-34.4e-12},pass-3'
    )
)
# produces:
[
    ("pass-1", {}),
    (
        "pass-2",
        {
            "arg1": 1,
            "arg2": "test",
            "arg3": "test-str",
            "arg-4": -3.44e-11,
        },
    ),
    ("pass-3", {}),
]
```

georgebisbas commented on tests/test_pass_lexer.py:
```diff
@@ -0,0 +1,104 @@
+
+def test_pass_parse_errors():
+    """
+    this test tries to trigger all parse errors in the parser.
```
Improve docstring?

AntonLydike replied:
Done!

georgebisbas reviewed (commented):
Would not something like this help to reduce all this code?
https://docs.python.org/3/library/argparse.html

AntonLydike commented:
> Would not something like this help to reduce all this code? https://docs.python.org/3/library/argparse.html

As far as I know argparse is not suited for parsing non-cli like strings (such as this). We use argparse to parse the actual command line args, but parsing the pass pipeline is (imo) a different beast and argparse doesn't seem to be suited for that...

mesham commented:
Very nice, thanks for adding this as I think it will be super helpful more widely for these things. I agree with your comments above, I don't think argparse can be easily used here unfortunately.

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+            raise PassPipelineParseError(name, "Expected pass name here")
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            yield name.span.text, dict()
+            return
+
+        if delim.kind is Kind.COMMA:
+            yield name.span.text, dict()
+            continue
+
+        if delim.kind is not Kind.L_BRACE:
+            raise PassPipelineParseError(
+                delim, "Expected a comma or pass arguments here"
+            )
```
Here, to simplify the logic, you could instead just check for `L_BRACE`, and return the empty dictionary instead?
Here you are assuming that the only next tokens can be `EOF` or `COMMA`, which shouldn't be the concern of this function?

AntonLydike replied:
I think I still want to raise an error in case the input is malformed, right? If someone makes an error an uses `pass1 pass2` it should tell it that the space there was unexpected.

math-fehr replied:
Well, you would get an error, though yes your error is `expected comma or pass arguments`, while the other way would give you `unexpected _`, so you would get a slightly worse error.
However, when maintaining code, this would mean that changing one rule might require to change other rules, as they reason about their position in the overall parser. I think this is the dangerous part, as it is very easy to miss that when you modify code, if you know what I mean.
Otherwise, I'm fine if you prefer it this way, I just find it a bit dangerous long-term.

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+
+        yield name.span.text, _parse_pass_args(lexer)
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            return
+        if delim.kind is not Kind.COMMA:
+            raise PassPipelineParseError(
+                delim, "Expected a comma after pass argument dict here"
+            )
```
I feel this logic should be separated from this function, and have a `parse_pipeline` and `_parse_pass`?
Idk how you feel about this?

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+            )
+
+
+def _parse_pass_args(lexer: PipelineLexer):
```
Maybe write here what is the syntax in the documentation?

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+            raise PassPipelineParseError(name, "Expected argument name here")
+
+        # handle zero-length args
+        if lexer.peek().kind in (Kind.SPACE, Kind.R_BRACE):
```
If you get a space here, you don't consume it, and I think you should?

math-fehr replied:
Maybe just `if `lexer.peek().kind` is not Kind.EQUALS:`?
I feel that you are often encoding the information that you know the next tokens, while actually you don't need to. If you just encode what you expect in your current rule, that is usually enough.

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+
+        args[name.span.text] = _parse_arg_value(lexer)
+
+        delim = lexer.lex()
```
Here as well, I would put the list logic outside of this function.

AntonLydike replied:
I couldn't figure out a nice way to do this as we need to either `return` or `continue`, depending on token

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+    return elms
+
+
+def _parse_arg_value_element(lexer: PipelineLexer) -> str | int | bool | float:
```
Maybe add an `_PassArgType` alias?

math-fehr replied:
So then if you add something to the list, it adds it to `_PassArgTypes` as well.

math-fehr reviewed (approved):
:+1:

superlopuh commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+                yield Token(Span(pos, pos + 1, input), Kind.EOF)
+                return
+
+    def lex(self) -> Token:
```
In my understanding, a lexer does everything in one pass, which would probably make sense here since we're never going to have command line arguments for pass pipelines that wouldn't make sense to store as a list. A list is also easier to debug and reason about since it forces the decoupling of the parsing and lexing stage.

AntonLydike replied:
I mean the lexer does it all in one pass, right?

superlopuh commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -0,0 +1,227 @@
+
+def parse_pipeline(
+    pipeline_spec: str,
+) -> Iterator[tuple[str, dict[str, _PassArgTypes]]]:
```
Seems like this could be a named tuple

AntonLydike replied:
Converted to dataclass!

AntonLydike commented:
@math-fehr, @superlopuh I cleaned up the code considerably, making use of the `match` clause and added lots of comments. I hope this made it clearer what the code does?

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -92,3 +93,175 @@ def peek(self) -> Token:
+@dataclass(eq=True, frozen=True)
+class PipelinePassSpec:
+    """
+    Small dataclass that holds the name and args of a specific pass.
```
```suggestion
    A pass name and its arguments.
```
I wouldn't say in documentation that a class is small, or that it is a dataclass, what do you think?

math-fehr commented on xdsl/utils/parse_pipeline.py:
```diff
@@ -92,3 +93,175 @@ def peek(self) -> Token:
+def _parse_pass_args(lexer: PipelineLexer) -> dict[str, _PassArgTypes]:
+    """
+    This parses pass arguments. They are a dictionary structure
+    with whitespace separated, multi-value elements:
```
Mention maybe that `{` is already parsed at that point.

math-fehr reviewed (approved):
+1
I still think the code is a bit complex, but as it is small enough I'm fine with it.

AntonLydike commented on tests/test_pass_lexer.py:
```diff
@@ -38,3 +43,93 @@ def test_pass_lex_errors():
+        ("pass-1,", ["pass-1"]),
+        ("pass-1{}", ["pass-1"]),
+        ("pass-1{arg1=true arg2}", ["pass-1"]),
+        ("pass-1{arg2 arg1=false}", ["pass-1"]),
```
I changed the test name and added a docstring, this is supposed to be testing edge cases in arg dict parsing.
